File: Backend/offers-service/src/main/java/com/smartek/offersservice/service/InterviewService.java
package com.smartek.offersservice.service;

import com.smartek.offersservice.dto.InterviewRequest;
import com.smartek.offersservice.dto.InterviewResponse;
import com.smartek.offersservice.entity.Application;
import com.smartek.offersservice.entity.Interview;
import com.smartek.offersservice.repository.ApplicationRepository;
import com.smartek.offersservice.repository.InterviewRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class InterviewService {
    
    private final InterviewRepository interviewRepository;
    private final ApplicationRepository applicationRepository;
    
    @Transactional
    public InterviewResponse createInterview(InterviewRequest request) {
        log.info("Creating interview for application ID: {}", request.getApplicationId());
        
        // Vérifier que la candidature existe
        Application application = applicationRepository.findById(request.getApplicationId())
                .orElseThrow(() -> new RuntimeException("Candidature non trouvée"));
        
        // Vérifier que la candidature est acceptée
        if (!"ACCEPTED".equals(application.getStatus())) {
            throw new RuntimeException("La candidature doit être acceptée pour planifier un entretien");
        }
        
        Interview interview = Interview.builder()
                .applicationId(request.getApplicationId())
                .offerId(application.getOfferId())
                .learnerId(application.getLearnerId())
                .learnerName(application.getLearnerName())
                .learnerEmail(application.getLearnerEmail())
                .interviewDate(request.getInterviewDate())
                .location(request.getLocation())
                .meetingLink(request.getMeetingLink())
                .notes(request.getNotes())
                .status(Interview.InterviewStatus.SCHEDULED)
                .createdBy(request.getCreatedBy())
                .build();
        
        Interview savedInterview = interviewRepository.save(interview);
        log.info("Interview created successfully with ID: {}", savedInterview.getId());
        
        return InterviewResponse.fromEntity(savedInterview);
    }
    
    public List<InterviewResponse> getInterviewsByOffer(Long offerId) {
        log.info("Fetching interviews for offer ID: {}", offerId);
        return interviewRepository.findByOfferId(offerId).stream()
                .map(InterviewResponse::fromEntity)
                .collect(Collectors.toList());
    }
    
    public List<InterviewResponse> getInterviewsByLearner(Long learnerId) {
        log.info("Fetching interviews for learner ID: {}", learnerId);
        return interviewRepository.findByLearnerId(learnerId).stream()
                .map(InterviewResponse::fromEntity)
                .collect(Collectors.toList());
    }
    
    public List<InterviewResponse> getInterviewsByApplication(Long applicationId) {
        log.info("Fetching interviews for application ID: {}", applicationId);
        return interviewRepository.findByApplicationId(applicationId).stream()
                .map(InterviewResponse::fromEntity)
                .collect(Collectors.toList());
    }
    
    public List<InterviewResponse> getAllInterviews() {
        log.info("Fetching all interviews");
        return interviewRepository.findAll().stream()
                .map(InterviewResponse::fromEntity)
                .collect(Collectors.toList());
    }
    
    @Transactional
    public InterviewResponse updateInterviewStatus(Long interviewId, String status) {
        log.info("Updating interview ID: {} to status: {}", interviewId, status);
        
        Interview interview = interviewRepository.findById(interviewId)
                .orElseThrow(() -> new RuntimeException("Entretien non trouvé"));
        
        interview.setStatus(Interview.InterviewStatus.valueOf(status));
        Interview updatedInterview = interviewRepository.save(interview);
        
        log.info("Interview status updated successfully");
        return InterviewResponse.fromEntity(updatedInterview);
    }
    
    @Transactional
    public InterviewResponse updateInterview(Long interviewId, InterviewRequest request) {
        log.info("Updating interview ID: {}", interviewId);
        
        Interview interview = interviewRepository.findById(interviewId)
                .orElseThrow(() -> new RuntimeException("Entretien non trouvé"));
        
        interview.setInterviewDate(request.getInterviewDate());
        interview.setLocation(request.getLocation());
        interview.setMeetingLink(request.getMeetingLink());
        interview.setNotes(request.getNotes());
        
        Interview updatedInterview = interviewRepository.save(interview);
        
        log.info("Interview updated successfully");
        return InterviewResponse.fromEntity(updatedInterview);
    }
    
    @Transactional
    public void deleteInterview(Long interviewId) {
        log.info("Deleting interview ID: {}", interviewId);
        interviewRepository.deleteById(interviewId);
        log.info("Interview deleted successfully");
    }
}

File: Frontend/angular-app/src/app/core/services/interview.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Interview, InterviewRequest } from '../models/interview.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class InterviewService {
  private apiUrl = `${environment.apiUrl}/interviews`;

  constructor(private http: HttpClient) {}

  createInterview(interview: InterviewRequest): Observable<Interview> {
    return this.http.post<Interview>(this.apiUrl, interview);
  }

  getAllInterviews(): Observable<Interview[]> {
    return this.http.get<Interview[]>(this.apiUrl);
  }

  getInterviewsByOffer(offerId: number): Observable<Interview[]> {
    return this.http.get<Interview[]>(`${this.apiUrl}/offer/${offerId}`);
  }

  getInterviewsByLearner(learnerId: number): Observable<Interview[]> {
    return this.http.get<Interview[]>(`${this.apiUrl}/learner/${learnerId}`);
  }

  getInterviewsByApplication(applicationId: number): Observable<Interview[]> {
    return this.http.get<Interview[]>(`${this.apiUrl}/application/${applicationId}`);
  }

  updateInterviewStatus(interviewId: number, status: string): Observable<Interview> {
    return this.http.put<Interview>(`${this.apiUrl}/${interviewId}/status?status=${status}`, {});
  }

  updateInterview(interviewId: number, interview: InterviewRequest): Observable<Interview> {
    return this.http.put<Interview>(`${this.apiUrl}/${interviewId}`, interview);
  }

  deleteInterview(interviewId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${interviewId}`);
  }
}
